enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,218 @@
+
+Actions:
+
+- Rework the `manifest` asset to generate a secret with CCM config.
+
+- Update unit tests.
+
+#### Build OpenStack CCM image by OCP automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy OpenStack CCM and test how it works.
+
+Actions:
+
+- Manually install CCM's [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OCP cluster deployed on OpenStack.
+
+- Update configuration of `kube-apiserver`, `kube-controller-manager` and `kubelet` by replacing `--cloud-provider openstack` with `--cloud-provider external` and removing `--cloud-config` parameters.
+
+**Note:** Example of manual testing: https://asciinema.org/a/303399?speed=2
+
+#### Write Cluster Cloud Controller Manager Operator
+
+The operator should be able to create, configure and manage different CCMs (including one for OpenStack) in OpenShift.
+
+Actions:
+
+- Create a new repo in OpenShift’s github: https://github.com/openshift/cluster-cloud-controller-manager-operator
+
+- Implement the operator, using [library-go](https://github.com/openshift/library-go) primitives.
+
+#### Build the operator image by OCP automation
+
+To start using CCM operator in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The image should be automatically tested before it becomes available. Dockerfile should be a part of the operator's repo.
+
+Actions:
+
+- Configure CI operator to build CCM operator image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OCP Quay account.
+
+#### Integrate the solution with OpenShift
+
+Actions:
+
+- Make sure Cinder CSI driver is supported in OpenShift.
+
+- Add CCM operator support in Cluster Version Operator. CCM operator will be deployed on early stages of installation and then it deploys and configures OpenStack CCM itself.
+
+- Change Kubelet configuration for OpenStack in [Machine Config Operator templates](https://github.com/openshift/machine-config-operator/blob/d044c74ea4b9900c269ee8de8131ed49ba6fedc8/templates/master/01-master-kubelet/openstack/units/kubelet.service.yaml#L30) to adopt external cloud provider.
```

Comment:
Should we have someone from CVO and MCO review?

Reply:
Yes, 3CMO (cluster cloud controller manager operator) will be managed by CVO so their opinion is really appreciated.

There won't be so much interaction with MCO, I believe. We will need to change just one line of code there:
https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-kubelet/openstack/units/kubelet.service.yaml#L30
But the more people look, the better :)